src/support/threads.cpp

```diff
@@ -0,0 +1,201 @@
+
+// Global thread information
+
+std::atomic_bool setMainThreadId(false);
+std::thread::id mainThreadId;
+
+struct MainThreadNoter {
+  MainThreadNoter() {
+    // global ctors are called on main thread
+    mainThreadId = std::this_thread::get_id();
+    setMainThreadId = true;
+  }
+};
+
+static MainThreadNoter mainThreadNoter;
+
+static std::unique_ptr<ThreadPool> pool;
+
+
+// Thread
+
+Thread::Thread() {
+  // main thread object's constructor itself can
+  // happen before onMainThread is ready
+  assert(onMainThread());
+  thread = std::unique_ptr<std::thread>(new std::thread(mainLoop, this));
+}
+
+Thread::~Thread() {
+  assert(onMainThread());
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    done = true;
+    condition.notify_one();
+  }
+  thread->join();
+}
+
+void Thread::work(std::function<ThreadWorkState ()> doWork_) {
+  // TODO: fancy work stealing
+  DEBUG_PRINT(std::cerr << "[Thread] send work to thread\n");
+  assert(onMainThread());
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    doWork = doWork_;
+    condition.notify_one();
+    DEBUG_PRINT(std::cerr << "[Thread] work sent\n");
+  }
+}
+
+bool Thread::onMainThread() {
+  // mainThread Id might not be set yet if we are in a global ctor, but
+  // that is on the main thread anyhow
+  return !setMainThreadId || std::this_thread::get_id() == mainThreadId;
+}
+
+void Thread::mainLoop(void *self_) {
+  auto* self = static_cast<Thread*>(self_);
+  while (1) {
+    DEBUG_PRINT(std::cerr << "[Thread] checking for work\n");
+    {
+      std::unique_lock<std::mutex> lock(self->mutex);
+      if (self->doWork) {
+        DEBUG_PRINT(std::cerr << "[Thread] doing work\n");
+        // run tasks until they are all done
+        while (self->doWork() == ThreadWorkState::More) {}
+        self->doWork = nullptr;
+      } else if (self->done) {
+        DEBUG_PRINT(std::cerr << "[Thread] done\n");
+        return;
+      }
+    }
+    ThreadPool::get()->notifyThreadIsReady();
+    {
+      std::unique_lock<std::mutex> lock(self->mutex);
+      if (!self->done && !self->doWork) {
+        DEBUG_PRINT(std::cerr << "[Thread] thread waiting\n");
+        self->condition.wait(lock);
+      }
+    }
+  }
+}
+
+
+// ThreadPool
+
+void ThreadPool::initialize(size_t num) {
+  if (num == 1) return; // no multiple cores, don't create threads
+  DEBUG_PRINT(std::cerr << "[Pool] initialize()\n");
+  std::unique_lock<std::mutex> lock(mutex);
+  ready.store(threads.size()); // initial state before first resetThreadsAreReady()
+  resetThreadsAreReady();
+  for (size_t i = 0; i < num; i++) {
+    threads.emplace_back(std::unique_ptr<Thread>(new Thread()));
+  }
+  DEBUG_PRINT(std::cerr << "[Pool] initialize() waiting\n");
+  condition.wait(lock, [this]() { return areThreadsReady(); });
+  DEBUG_PRINT(std::cerr << "[Pool] initialize() is done\n");
+}
+
+ThreadPool* ThreadPool::get() {
+  if (!pool) {
+    assert(Thread::onMainThread());
+    size_t num = std::thread::hardware_concurrency();
```

Comment:
`size_t num = std::max(1, std::thread::hardware_concurrency());`
and drop the next line.